File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/db/BoardTableConnection.java
/*
 * Copyright (c) 2016. Eli Connelly
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.emogoth.android.phone.mimi.db;


import android.content.ContentValues;
import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.text.TextUtils;
import android.util.Log;

import com.emogoth.android.phone.mimi.R;
import com.emogoth.android.phone.mimi.app.MimiApplication;
import com.emogoth.android.phone.mimi.db.model.BaseModel;
import com.emogoth.android.phone.mimi.db.model.Board;
import com.mimireader.chanlib.models.ChanBoard;
import com.squareup.sqlbrite3.BriteDatabase;

import org.reactivestreams.Publisher;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.Callable;

import io.reactivex.Flowable;
import io.reactivex.Single;
import io.reactivex.SingleSource;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;

public class BoardTableConnection {
    public static final String LOG_TAG = BoardTableConnection.class.getSimpleName();

    public static Single<List<Board>> fetchBoards(int orderBy) {
        return DatabaseUtils.fetchTable(Board.class, Board.TABLE_NAME, Board.sortOrder(orderBy), Board.KEY_VISIBLE + "=?", true);
    }

    public static Flowable<List<ChanBoard>> observeBoards(int orderBy) {
        return DatabaseUtils.observeTable(Board.class, Board.TABLE_NAME, Board.sortOrder(orderBy), Board.KEY_VISIBLE + "=?", true)
                .flatMap((Function<List<Board>, Publisher<List<ChanBoard>>>) boards -> Flowable.just(convertBoardDbModelsToChanBoards(boards)));
    }

    public static Single<ChanBoard> fetchBoard(final String boardName) {
        if (TextUtils.isEmpty(boardName)) {
            return Single.just(new ChanBoard());
        }
        return DatabaseUtils.fetchTable(Board.class, Board.TABLE_NAME, null, Board.KEY_NAME + "=?", boardName)
                .flatMap((Function<List<Board>, Single<ChanBoard>>) boards -> {
                    if (boards != null && boards.size() > 0) {
                        return Single.just(convertBoardDbModelToBoard(boards.get(0)));
                    }
                    return Single.just(new ChanBoard());
                });
    }

    public static Single<Boolean> setBoardFavorite(final String boardName, final boolean favorite) {
        Board b = new Board();
        b.name = boardName;

        Board.Builder builder = new Board.Builder();
        builder.favorite(favorite);

        b.setValues(builder.build());

        return DatabaseUtils.updateTable(b);
    }

    public static Single<Boolean> incrementAccessCount(final String boardName) {
        return updateBoard(boardName, 2);
    }

    public static Single<Boolean> incrementPostCount(final String boardName) {
        return updateBoard(boardName, 1);
    }

    public static Single<Boolean> updateLastAccess(final String boardName) {
        return updateBoard(boardName, 0);
    }

    private static Single<Boolean> updateBoard(final String boardName, final int type) {
        return DatabaseUtils.fetchTable(Board.class, Board.TABLE_NAME, null, Board.KEY_NAME + "=?", boardName)
                .flatMap((Function<List<Board>, Single<Boolean>>) boards -> {
                    if (boards != null && boards.size() > 0) {
                        Board board = boards.get(0);
                        ContentValues values = null;
                        if (type == 0) {
                            values = updateLastAccessed();
                        } else if (type == 1) {
                            values = updatePostCount(board);
                        } else if (type == 2) {
                            values = updateAccessCount(board);
                        }

                        if (values != null) {
                            board.setValues(values);
                            return DatabaseUtils.updateTable(board);
                        }
                    }

                    return Single.just(false);
                });
    }

    private static ContentValues updateLastAccessed() {
        ContentValues values = new ContentValues();
        values.put(Board.KEY_LAST_ACCESSED, System.currentTimeMillis());

        return values;
    }

    private static ContentValues updatePostCount(Board board) {
        int count = board.postCount == 0 ? 1 : board.postCount + 1;

        ContentValues values = new ContentValues();
        values.put(Board.KEY_POST_COUNT, count);

        return values;
    }

    private static ContentValues updateAccessCount(Board board) {
        int count = board.accessCount == 0 ? 1 : board.accessCount + 1;

        ContentValues values = new ContentValues();
        values.put(Board.KEY_ACCESS_COUNT, count);

        return values;
    }

    public static Single<Boolean> resetStats() {
        return Single.defer((Callable<SingleSource<Boolean>>) () -> {
            BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
            BriteDatabase.Transaction transaction = db.newTransaction();

            ContentValues values = new ContentValues();

            values.put(Board.KEY_ACCESS_COUNT, 0);
            values.put(Board.KEY_LAST_ACCESSED, 0);
            values.put(Board.KEY_POST_COUNT, 0);

            int val = 0;
            try {
                val = db.update(Board.TABLE_NAME, SQLiteDatabase.CONFLICT_IGNORE, values, null, null);
                transaction.markSuccessful();
            } catch (Exception e) {
                Log.e(LOG_TAG, "Error putting post options into the database", e);
            } finally {
                transaction.end();
            }

            return Single.just(val > 0);
        });

    }

    public static Flowable<Boolean> updateBoardOrder(final List<ChanBoard> boards) {
        return Flowable.defer((Callable<Flowable<Boolean>>) () -> {
            int val;
            boolean success = false;
            BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
            BriteDatabase.Transaction transaction = db.newTransaction();
            ContentValues values = new ContentValues();

            try {
                for (int i = 0; i < boards.size(); i++) {
                    values.put(Board.KEY_ORDER_INDEX, i);
                    val = db.update(Board.TABLE_NAME, SQLiteDatabase.CONFLICT_REPLACE, values, Board.KEY_NAME + "=?", boards.get(i).getName());

                    success = val > 0;
                    values.clear();
                }

                transaction.markSuccessful();
            } catch (Exception e) {
                Log.e(LOG_TAG, "Error updating board order", e);
            } finally {
                transaction.end();
                return Flowable.just(success).onErrorReturn(throwable -> {
                    Log.e(LOG_TAG, "Error updating board order", throwable);
                    return false;
                });
            }

        })
                .compose(DatabaseUtils.applySchedulers());
    }

    public static ChanBoard convertBoardDbModelToBoard(final Board oldBoard) {

        final ChanBoard newBoard = new ChanBoard();
        final String boardName = oldBoard.name; // Path is now Name
        final String boardTitle = oldBoard.title; // Name is now Title
        int ws = 1;
        if (oldBoard.nsfw == 1) {
            ws = 0;
        }
        final boolean favorite = oldBoard.favorite == 1;
        final int postsPerPage = oldBoard.perPage;
        final int pages = oldBoard.pages;

        newBoard.setName(boardName);
        newBoard.setTitle(boardTitle);
        newBoard.setWsBoard(ws);
        newBoard.setFavorite(favorite);
        newBoard.setPerPage(postsPerPage);
        newBoard.setPages(pages);

        return newBoard;
    }

    public static List<ChanBoard> convertBoardDbModelsToChanBoards(final List<Board> boards) {
        if (boards == null || boards.size() == 0) {
            return Collections.emptyList();
        }

        List<ChanBoard> chanBoards = new ArrayList<>(boards.size());
        for (Board board : boards) {
            chanBoards.add(convertBoardDbModelToBoard(board));
        }

        return chanBoards;
    }

    public static Board convertChanBoardToDbModel(ChanBoard chanBoard) {
        Board board = new Board();
        board.title = chanBoard.getTitle();
        board.name = chanBoard.getName();
        board.favorite = (byte) (chanBoard.isFavorite() ? 1 : 0);
        board.nsfw = (byte) (chanBoard.getWsBoard() == 1 ? 1 : 0);
        board.pages = chanBoard.getPages();
        board.perPage = chanBoard.getPerPage();
        board.maxFileSize = chanBoard.getMaxFilesize();

        return board;
    }

    public static List<ChanBoard> filterVisibleBoards(Context context, List<ChanBoard> boards) {
        final ArrayList<String> visibleBoardNames = new ArrayList<>();
        if (boards != null && boards.size() > 0) {
            final String[] boardsArray = context.getResources().getStringArray(R.array.boards);

            for (int i = 0; i < boardsArray.length; i++) {
                visibleBoardNames.add(boardsArray[i].replaceAll("/", ""));
            }

            List<ChanBoard> filteredBoards = new ArrayList<>(boards.size());
            for (ChanBoard board : boards) {
                if (visibleBoardNames.indexOf(board.getName()) >= 0) {
                    filteredBoards.add(board);
                }
            }


            return filteredBoards;

        }

        return Collections.emptyList();
    }

    public static Single<List<ChanBoard>> initDefaultBoards(final Context context) {
        return Single
                .defer((Callable<Single<String[]>>) () -> Single.just(context.getResources().getStringArray(R.array.boards)))
                .map(strings -> {
                    List<ChanBoard> boards = new ArrayList<>(strings.length);
                    for (String boardName : strings) {
                        ChanBoard board = new ChanBoard();
                        board.setName(boardName);
                        board.setTitle("");
                        board.setWsBoard(1);
                        boards.add(board);
                    }
                    return boards;
                })
                .doOnSuccess(BoardTableConnection.saveBoards())
                .toFlowable()
                .flatMapIterable((Function<List<ChanBoard>, Iterable<ChanBoard>>) board -> board)
                .flatMap((Function<ChanBoard, Flowable<ChanBoard>>) s -> setBoardVisibility(s.getName(), true).toFlowable())
                .toList();
    }

    public static Consumer<? extends BaseModel> saveToDatabase() {
        return (Consumer<BaseModel>) baseModel -> {
            BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
            BriteDatabase.Transaction transaction = db.newTransaction();
            try {
                DatabaseUtils.update(db, baseModel);
                transaction.markSuccessful();
            } catch (Exception e) {
                Log.e(LOG_TAG, "Database update failed", e);
            } finally {
                transaction.end();
            }
        };
    }

    public static Consumer<List<ChanBoard>> saveBoards() {
        return BoardTableConnection::saveBoards;
    }

    public static void saveBoards(List<ChanBoard> chanBoards) {
        if (chanBoards == null || chanBoards.size() == 0) {
            return;
        }

        BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
        BriteDatabase.Transaction transaction = db.newTransaction();

        try {
            for (ChanBoard board : chanBoards) {
                int updateVal = DatabaseUtils.update(db, convertChanBoardToDbModel(board));

                if (updateVal == 0) {
                    DatabaseUtils.insert(db, convertChanBoardToDbModel(board));
                }
            }
            transaction.markSuccessful();
        } catch (Exception e) {
            Log.e(LOG_TAG, "Database update failed", e);
        } finally {
            transaction.end();
        }
    }

    public static Single<ChanBoard> setBoardVisibility(final String boardPath, final Boolean visible) {
        if (TextUtils.isEmpty(boardPath)) {
            return Single.just(new ChanBoard());
        }

        return Single.defer((Callable<Single<Boolean>>) () -> {
            BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
            BriteDatabase.Transaction transaction = db.newTransaction();

            int val = 0;

            try {
                ContentValues contentValues = new ContentValues();
                contentValues.put(Board.KEY_VISIBLE, visible ? 1 : 0);

                String name = boardPath.replaceAll("/", "");
                val = db.update(Board.TABLE_NAME, SQLiteDatabase.CONFLICT_REPLACE, contentValues, Board.KEY_NAME + "=?", name);

                transaction.markSuccessful();
            } catch (Exception e) {
                Log.e(LOG_TAG, "Database update failed", e);
            } finally {
                transaction.end();
            }

            return Single.just(val > 0);
        })
                .flatMap((Function<Boolean, Single<ChanBoard>>) success -> {
                    if (success) {
                        return fetchBoard(boardPath);
                    }

                    return Single.just(new ChanBoard());
                })
                .compose(DatabaseUtils.applySingleSchedulers());
    }

}

File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/db/BoardTableConnection.kt
package com.emogoth.android.phone.mimi.db

import android.content.Context
import android.util.Log
import com.emogoth.android.phone.mimi.R
import com.emogoth.android.phone.mimi.db.models.Board
import com.mimireader.chanlib.models.ChanBoard
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.functions.Consumer
import kotlin.collections.ArrayList

object BoardTableConnection {
    const val LOG_TAG = "BoardTableConnection"

    @JvmStatic
    fun fetchBoards(orderBy: Int): Single<List<Board>> {
        val db = MimiDatabase.getInstance()?.boards() ?: return Single.just(emptyList())

        val boardsObs: Flowable<List<Board>> =
                when (orderBy) {
                    0 -> db.getAllOrderByName()
                    1 -> db.getAllOrderByTitle()
                    2 -> db.getAllOrderByName()
                    3 -> db.getAllOrderByAccessCount()
                    4 -> db.getAllOrderByPostCount()
                    5 -> db.getAllOrderByLastAccessed()
                    6 -> db.getAllOrderByFavorite()
                    7 -> db.getAllOrderByCustom()
                    else -> db.getAllOrderByName()
                }
        return boardsObs.first(emptyList())
    }

    @JvmStatic
    fun observeBoards(orderBy: Int): Flowable<List<ChanBoard>> {
        val db = MimiDatabase.getInstance()?.boards()
                ?: return Flowable.just(emptyList())

        val boardsObs: Flowable<List<Board>> =
                when (orderBy) {
                    0 -> db.getAllOrderByName()
                    1 -> db.getAllOrderByTitle()
                    2 -> db.getAllOrderByName()
                    3 -> db.getAllOrderByAccessCount()
                    4 -> db.getAllOrderByPostCount()
                    5 -> db.getAllOrderByLastAccessed()
                    6 -> db.getAllOrderByFavorite()
                    7 -> db.getAllOrderByCustom()
                    else -> db.getAllOrderByName()
                }
        return boardsObs.flatMap { Flowable.just(convertBoardDbModelsToChanBoards(it)) }
    }

    @JvmStatic
    fun fetchBoard(boardName: String): Single<ChanBoard> {
        return if (boardName.isEmpty()) {
            Single.just(ChanBoard())
        } else {
            MimiDatabase.getInstance()?.boards()?.getBoard(boardName)?.onErrorReturn { Board() }?.flatMap { Single.just(convertBoardDbModelToBoard(it)) }
                    ?: Single.just(ChanBoard())
        }

    }

    @JvmStatic
    fun setBoardFavorite(boardName: String, favorite: Boolean): Single<Boolean> {
        return MimiDatabase.getInstance()?.boards()?.getBoard(boardName)?.onErrorReturn { Board() }?.flatMap { board ->
            if (board.id == null) {
                Single.just(false)
            } else {
                val b = Board(
                        id = board.id,
                        title = board.title,
                        name = board.name,
                        favorite = favorite,
                        visible = board.visible,
                        nsfw = board.nsfw,
                        pages = board.pages,
                        perPage = board.perPage,
                        maxFileSize = board.maxFileSize,
                        accessCount = board.accessCount,
                        postCount = board.postCount,
                        lastAccessed = board.lastAccessed,
                        orderIndex = board.orderIndex,
                        category = board.category)

                val updated = MimiDatabase.getInstance()?.boards()?.updateBoard(b) ?: 0 > 0
                Single.just(updated)
            }
        } ?: Single.just(false)
    }

    @JvmStatic
    fun incrementAccessCount(boardName: String): Single<Boolean> {
        return MimiDatabase.getInstance()?.boards()?.getBoard(boardName)?.onErrorReturn { Board() }?.flatMap { board ->
            if (board.id == null) {
                Single.just(false)
            } else {
                val b = Board(
                        id = board.id,
                        title = board.title,
                        name = board.name,
                        favorite = board.favorite,
                        visible = board.visible,
                        nsfw = board.nsfw,
                        pages = board.pages,
                        perPage = board.perPage,
                        maxFileSize = board.maxFileSize,
                        accessCount = board.accessCount + 1,
                        postCount = board.postCount,
                        lastAccessed = board.lastAccessed,
                        orderIndex = board.orderIndex,
                        category = board.category)

                val updated = MimiDatabase.getInstance()?.boards()?.updateBoard(b) ?: 0 > 0
                Single.just(updated)
            }
        } ?: Single.just(false)
    }

    @JvmStatic
    fun incrementPostCount(boardName: String): Single<Boolean> {
        return MimiDatabase.getInstance()?.boards()?.getBoard(boardName)?.onErrorReturn { Board() }?.flatMap { board ->
            if (board.id == null) {
                Single.just(false)
            } else {
                val b = Board(
                        id = board.id,
                        title = board.title,
                        name = board.name,
                        favorite = board.favorite,
                        visible = board.visible,
                        nsfw = board.nsfw,
                        pages = board.pages,
                        perPage = board.perPage,
                        maxFileSize = board.maxFileSize,
                        accessCount = board.accessCount,
                        postCount = board.postCount + 1,
                        lastAccessed = board.lastAccessed,
                        orderIndex = board.orderIndex,
                        category = board.category)

                val updated = MimiDatabase.getInstance()?.boards()?.updateBoard(b) ?: 0 > 0
                Single.just(updated)
            }
        } ?: Single.just(false)
    }

    @JvmStatic
    fun updateLastAccess(boardName: String): Single<Boolean> {
        return MimiDatabase.getInstance()?.boards()?.getBoard(boardName)?.onErrorReturn { Board() }?.flatMap { board ->
            if (board.id == null) {
                Single.just(false)
            } else {
                val b = Board(
                        id = board.id,
                        title = board.title,
                        name = board.name,
                        favorite = board.favorite,
                        visible = board.visible,
                        nsfw = board.nsfw,
                        pages = board.pages,
                        perPage = board.perPage,
                        maxFileSize = board.maxFileSize,
                        accessCount = board.accessCount,
                        postCount = board.postCount,
                        lastAccessed = System.currentTimeMillis(),
                        orderIndex = board.orderIndex,
                        category = board.category)

                val updated = MimiDatabase.getInstance()?.boards()?.updateBoard(b) ?: 0 > 0
                Single.just(updated)
            }
        } ?: Single.just(false)
    }

    @JvmStatic
    fun resetStats(): Single<Boolean> {
        return MimiDatabase.getInstance()?.boards()?.getVisibleBoards()?.flatMapIterable {
            val allBoards = ArrayList<Board>()
            allBoards.addAll(it)
            allBoards
        }?.single(Board())
                ?.flatMap { board ->
                    val b = Board(
                            id = board.id,
                            title = board.title,
                            name = board.name,
                            favorite = board.favorite,
                            visible = board.visible,
                            nsfw = board.nsfw,
                            pages = board.pages,
                            perPage = board.perPage,
                            maxFileSize = board.maxFileSize,
                            accessCount = 0,
                            postCount = 0,
                            lastAccessed = 0,
                            orderIndex = board.orderIndex,
                            category = board.category)

                    val updated = MimiDatabase.getInstance()?.boards()?.updateBoard(b) ?: 0 > 0
                    Single.just(updated)
                } ?: Single.just(false)
    }

    @JvmStatic
    fun updateBoardOrder(boards: List<ChanBoard>): Single<Boolean> {
        return MimiDatabase.getInstance()?.boards()?.getVisibleBoards()?.first(emptyList())?.flatMap { boardList ->

            val orderedList = ArrayList<Board>(boards.size)
            for (i in boards.indices) {
                for (j in boardList.indices) {
                    val board = boardList[j]
                    if (boards[i].name == board.name) {
                        val b = Board(
                                id = board.id,
                                title = board.title,
                                name = board.name,
                                favorite = board.favorite,
                                visible = board.visible,
                                nsfw = board.nsfw,
                                pages = board.pages,
                                perPage = board.perPage,
                                maxFileSize = board.maxFileSize,
                                accessCount = board.accessCount,
                                postCount = board.postCount,
                                lastAccessed = board.lastAccessed,
                                orderIndex = i,
                                category = board.category)

                        orderedList.add(b)
                        break
                    }
                }
            }

            Single.just(orderedList)
        }?.flatMap {
            val updated = MimiDatabase.getInstance()?.boards()?.update(it) ?: 0 > 0
            Single.just(updated)
        } ?: Single.just(false)
    }

    fun convertBoardDbModelToBoard(oldBoard: Board): ChanBoard {
        if (oldBoard.id == null) {
            return ChanBoard()
        }

        val newBoard = ChanBoard()
        val boardName = oldBoard.name // Path is now Name
        val boardTitle = oldBoard.title // Name is now Title
        val ws = if (oldBoard.nsfw) 0 else 1
        val favorite = oldBoard.favorite
        val postsPerPage = oldBoard.perPage
        val pages = oldBoard.pages
        newBoard.name = boardName
        newBoard.title = boardTitle
        newBoard.wsBoard = ws
        newBoard.isFavorite = favorite
        newBoard.perPage = postsPerPage
        newBoard.pages = pages
        return newBoard
    }

    @JvmStatic
    fun convertBoardDbModelsToChanBoards(boards: List<Board>): List<ChanBoard> {
        if (boards.isEmpty()) {
            return emptyList()
        }

        val chanBoards = ArrayList<ChanBoard>(boards.size)
        for (board in boards) {
            val b = convertBoardDbModelToBoard(board)
            if (!b.isEmpty) {
                chanBoards.add(b)
            }
        }
        return chanBoards
    }

    private fun convertChanBoardToDbModel(chanBoard: ChanBoard): Board {
        return Board(title = chanBoard.title, name = chanBoard.name, favorite = chanBoard.isFavorite, nsfw = chanBoard.wsBoard == 0, pages = chanBoard.pages, perPage = chanBoard.perPage, maxFileSize = chanBoard.maxFilesize)
    }

    private fun convertChanBoardListToDbModel(chanBoards: List<ChanBoard>): List<Board> {
        val boards = ArrayList<Board>(chanBoards.size)
        for (cb in chanBoards) {
            boards.add(convertChanBoardToDbModel(cb))
        }

        return boards
    }

    @JvmStatic
    fun filterVisibleBoards(context: Context, boards: List<ChanBoard>): List<ChanBoard> {
        val visibleBoardNames = ArrayList<String>()
        if (boards.isNotEmpty()) {
            val boardsArray = context.resources.getStringArray(R.array.boards)
            for (i in boardsArray.indices) {
                visibleBoardNames.add(boardsArray[i].replace("/".toRegex(), ""))
            }
            val filteredBoards: MutableList<ChanBoard> = ArrayList(boards.size)
            for (board in boards) {
                if (visibleBoardNames.indexOf(board.name) >= 0) {
                    filteredBoards.add(board)
                }
            }
            return filteredBoards
        }
        return emptyList()
    }

    @JvmStatic
    fun initDefaultBoards(context: Context): Single<List<ChanBoard>> {
        return Single
                .defer { Single.just(context.resources.getStringArray(R.array.boards)) }
                .map { strings: Array<String> ->
                    val boards: MutableList<ChanBoard> = ArrayList(strings.size)
                    for (boardName in strings) {
                        val board = ChanBoard()
                        board.name = boardName
                        board.title = ""
                        board.wsBoard = 1
                        boards.add(board)
                    }
                    boards
                }
                .doOnSuccess(saveBoards())
                .toFlowable()
                .flatMapIterable { board: List<ChanBoard> -> board }
                .flatMap { s: ChanBoard -> setBoardVisibility(s, true).toFlowable() }
                .toList()
    }

    @JvmStatic
    fun saveBoards(): Consumer<List<ChanBoard>> {
        return Consumer { saveBoards(it) }
    }

    @JvmStatic
    fun saveBoards(chanBoards: List<ChanBoard>) {
        Log.d(LOG_TAG, "Calling saveBoards()")
        if (chanBoards.isEmpty()) {
            return
        }

        try {
            val boards = convertChanBoardListToDbModel(chanBoards)

            MimiDatabase.getInstance()?.boards()?.upsert(boards)
        } catch (e: Exception) {
            Log.e(LOG_TAG, "Error while inserting boards", e)
        }
    }

    @JvmStatic
    fun setBoardVisibility(board: ChanBoard, visible: Boolean): Single<ChanBoard> {
        Log.d(LOG_TAG, "Calling setBoardVisibility(${board.name}, $visible)")
        return MimiDatabase.getInstance()?.boards()?.getBoard(board.name)?.onErrorReturn { Board() }?.flatMap { bd ->
            if (bd.id == null) {
                Single.just(ChanBoard())
            } else {
                val b = Board(
                        id = bd.id,
                        title = bd.title,
                        name = bd.name,
                        favorite = bd.favorite,
                        visible = visible,
                        nsfw = bd.nsfw,
                        pages = bd.pages,
                        perPage = bd.perPage,
                        maxFileSize = bd.maxFileSize,
                        accessCount = bd.accessCount,
                        postCount = bd.postCount,
                        lastAccessed = bd.lastAccessed,
                        orderIndex = bd.orderIndex,
                        category = bd.category)
                if (bd.visible != visible) {
                    val updated = MimiDatabase.getInstance()?.boards()?.updateBoard(b) ?: 0 > 0
                }
                Single.just(convertBoardDbModelToBoard(b))
            }
        } ?: Single.just(ChanBoard())
    }
}